src/libmeasurement_kit/dns/system_resolver.hpp

```diff
@@ -0,0 +1,141 @@
+                    Var<Logger> l) {
+        dns_class = dns_c;
+        dns_type = dns_t;
+        name = n;
+        cb = c;
+        settings = s;
+        reactor = r;
+        logger = l;
+        memset(&hints, 0, sizeof(hints));
+    }
+
+    ~ResolverContext() {
+        if (servinfo != nullptr) {
+            freeaddrinfo(servinfo);
+        }
+    }
+};
+
+template<MK_MOCK(getaddrinfo), MK_MOCK(inet_ntop)>
+static inline void resolve_async(ResolverContext *context) {
+    std::unique_ptr<ResolverContext> ctx(context);
+    Callback<Error, Var<Message>> callback = ctx->cb;
+    Var<Message> message_copy;
+
+    int error = getaddrinfo(ctx->name.c_str(), nullptr, &(ctx->hints),
+                            &(ctx->servinfo));
+    if (error) {
+        // check the error variable and return the correct error
+        ctx->logger->warn(gai_strerror(error));
+        ctx->reactor->call_soon(
+            [callback]() { callback(ResolverError(), nullptr); });
+        return;
+    }
+
+    assert(ctx->servinfo != nullptr);
+
+    void *addr_ptr;
+    char address[128];
+    std::vector<Answer> answers;
+    for (addrinfo *p = ctx->servinfo; p != nullptr; p = p->ai_next) {
+        Answer answer;
+        answer.name = ctx->name;
+        answer.qclass = ctx->dns_class;
+        if (p->ai_family == AF_INET) {
+            answer.type = QueryTypeId::A;
+            addr_ptr = &((sockaddr_in *)p->ai_addr)->sin_addr;
+        } else if (p->ai_family == AF_INET6) {
+            answer.type = QueryTypeId::AAAA;
+            addr_ptr = &((sockaddr_in6 *)p->ai_addr)->sin6_addr;
+        } else {
+            ctx->reactor->call_soon(
+                [callback]() { callback(ResolverError(), nullptr); });
+            return;
+        }
+        if (inet_ntop(p->ai_family, addr_ptr, address, sizeof(address)) ==
+            NULL) {
+            ctx->logger->warn("dns: unexpected inet_ntop failure");
+            throw std::runtime_error("Unexpected inet_ntop failure");
+        }
+        if (p->ai_family == AF_INET) {
+            answer.ipv4 = std::string(address);
+        } else if (p->ai_family == AF_INET6) {
+            answer.ipv6 = std::string(address);
+        }
+        answers.push_back(answer);
+    }
+    ctx->message->answers = answers;
+    message_copy = ctx->message;
+    ctx->reactor->call_soon(
+        [callback, message_copy]() { callback(NoError(), message_copy); });
+}
+
+template<MK_MOCK(getaddrinfo), MK_MOCK(inet_ntop)>
+inline void system_resolver(QueryClass dns_class, QueryType dns_type,
+                            std::string name, Callback<Error, Var<Message>> cb,
+                            Settings settings, Var<Reactor> reactor,
+                            Var<Logger> logger) {
+    std::unique_ptr<ResolverContext> ctx(new ResolverContext(dns_class, dns_type, name, cb,
+                                               settings, reactor, logger));
+    Query query;
+    ctx->hints.ai_flags = AI_ALL;
```

Comment:
From `getaddrinfo(3)` on macOS:

```
    The AI_ALL bit without the AI_V4MAPPED bit is ignored.
```

Here instead I'd specify `AI_CANONNAME` and will also get and store it in the response (I recall this is a piece of information to which @hellais was interested and now I see we also have issue #742 for that)